dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/PreCheckSecurityOwnershipValidationHook.java

```diff
@@ -276,37 +303,201 @@ private String getTeiUidFromEvent( TrackerImportValidationContext context, Event
+        String trackedEntity, boolean isCreatableInSearchScope )
+    {
+        TrackerBundle bundle = reporter.getValidationContext().getBundle();
+        User user = bundle.getUser();
+
+        checkNotNull( user, USER_CANT_BE_NULL );
+        checkNotNull( programStage, PROGRAM_STAGE_CANT_BE_NULL );
+        checkNotNull( programStage.getProgram(), PROGRAM_CANT_BE_NULL );
+        checkNotNull( eventOrgUnit, ORGANISATION_UNIT_CANT_BE_NULL );
+
+        if ( isCreatableInSearchScope ? !organisationUnitService.isInUserSearchHierarchyCached( user, eventOrgUnit )
+            : !organisationUnitService.isInUserHierarchyCached( user, eventOrgUnit ) )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1000 )
+                .addArg( user )
+                .addArg( eventOrgUnit ) );
+        }
+
+        if ( programStage.getProgram().isWithoutRegistration() )
+        {
+            checkProgramWriteAccess( reporter, user, programStage.getProgram() );
+        }
+        else
+        {
+            checkProgramStageWriteAccess( reporter, user, programStage );
+            final Program program = programStage.getProgram();
+
+            checkProgramReadAccess( reporter, user, program );
+
+            checkTeiTypeAndTeiProgramAccess( reporter, user,
+                trackedEntity,
+                ownerOrgUnit,
+                programStage.getProgram() );
+        }
+
+        if ( categoryOptionCombo != null )
+        {
+            checkWriteCategoryOptionComboAccess( reporter, categoryOptionCombo );
+        }
+    }
+
+    private void checkProgramReadAccess( ValidationErrorReporter reporter, User user,
+        Program program )
+    {
+        checkNotNull( user, USER_CANT_BE_NULL );
+        checkNotNull( program, PROGRAM_CANT_BE_NULL );
+
+        if ( !aclService.canDataRead( user, program ) )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1096 )
+                .addArg( user )
+                .addArg( program ) );
+        }
+    }
+
+    private void checkProgramStageWriteAccess( ValidationErrorReporter reporter, User user,
+        ProgramStage programStage )
+    {
+        checkNotNull( user, USER_CANT_BE_NULL );
+        checkNotNull( programStage, PROGRAM_STAGE_CANT_BE_NULL );
+
+        if ( !aclService.canDataWrite( user, programStage ) )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1095 )
+                .addArg( user )
+                .addArg( programStage ) );
+        }
+    }
+
+    private void checkProgramWriteAccess( ValidationErrorReporter reporter, User user,
+        Program program )
+    {
+        checkNotNull( user, USER_CANT_BE_NULL );
+        checkNotNull( program, PROGRAM_CANT_BE_NULL );
+
+        if ( !aclService.canDataWrite( user, program ) )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1091 )
+                .addArg( user )
+                .addArg( program ) );
+        }
+    }
```

Comment:
These look very similar ;)

Reply:
true 😄 this entire class can be refactored at a later point. I have lots of things to move around and test first. main goal is to untangle the report creation and remove the merge logic.